### ISSUE TITLE
BlockClosure's DNU vs. JavaScript aware BlockClosures that might not

### ISSUE THREAD

**Issue by sebastianconcept** [amber-smalltalk/amber#1130]
Currently one cannot do `window require undef: 'something'` using Smalltalk.

Of course is not just about RequireJS and that `undef` (sub)function, there is a myriad of libraries that has functions that can return other functions.

So, right now require returns a BlockClosure, it DNU's and the show stops.

With this feature request, that show could go on.

I think Amber would be better if its BlockClosures could be just a bit more reflective than they are today and before DNU'ing, they can look at their js object (with a `^ basicAt: #selector ifAbsent: [ super dnu]` to see if they have actually something there and, if so, return it. 

If we do this feature, Amber could interact with the js world even more naturally.

What do you think?

**Comment by herby** [comment]
There is `allowJavaScript` for that. Just use it. Marking won't fix.

If you want to have it automatic for all BlickClosures, make an issue for that (or rename and rephrase and teopen this).

**Comment by herby** [comment]
But, also consider #1116. I in fact want to remove `allowJavaScriptCall` (and so, this "smart" fallback from Smalltalk BlockClosure to methods appended as API in a concrete instance) completely, and be explicit where the thing is to be treated as BlockClosure or as a plain JS object.

**Comment by sebastianconcept** [comment]
Ah so o #1116 will provide this enhancement @herby ?

**Comment by herby** [comment]
??? No, #1116 wants to remove it completely. Not it is present, opt-in as `.allowJavaScript|Calls = true`. As is shown in #1116 description.

Sebastian Sastre wrote:

**Comment by herby** [comment]
That is, #1116 wants to remove the "dual-nature, ST first with fallback to JS". You can still call the JS API of an object you `Smalltalk optOut:`ed.

Sebastian Sastre wrote:

**Comment by sebastianconcept** [comment]
How would you write this:
`window.require.undef('something')`

in Amber after #1116 gets released?

**Comment by herby** [comment]
Sebastian Sastre wrote:
> How would you write this:
> |window.require.undef('something')|

As always, `require undef: 'something'. But you then won't be able to do `require value: 'module-name'`, as `value:` is Smalltalk api, you would need to call it via direct call using `NativeFunction of: require value: 'module-name'`.

For functions that are more API holder (like jQuery) this works great (and is already used, `allowJavaScriptCalls` was replaced by #1116. For functions that want to be callable but at the same time be API holders, #1116 makes things harder (as shown above).

> in Amber after #1116 
> <https://github.com/amber-smalltalk/amber/issues/1116> gets released?

The API is already there, in 0.14.1. `Smalltalk optOut: require`, for this case.
What is not out, is deprecation of `allowJavaScriptCalls`, as there is not a discussion about explicit nature (ST by default, optOut:able to JS) versus opt-in fallback to JS models.

**Comment by sebastianconcept** [comment]
sounds good.

We could also make the minus one word version: `require asNativeFuction value: 'module-name'`

**Comment by herby** [comment]
We can't, as we have  nothing obvious which `asNativeFunction` would return - the object lost its ST nature by optOut: message. The trick that it works lies exactly on it that it is done within one step, in one selector `of:value:`, implemented in inline JS (there is whole family of them there). You can of course make a wrapper object that forwards value: family to NativeFunction of:value: family, but it isn't something I would do, if you want, you can.
